winrm/protocol.py

```diff
@@ -395,6 +395,31 @@ def cleanup_command(self, shell_id, command_id):
         # TODO change assert into user-friendly exception
         assert uuid.UUID(relates_to.replace('uuid:', '')) == message_id
 
+    def send_command_input(self, shell_id, command_id, stdin_input):
+        """
+        Send input to the given shell and command.
+        @param string shell_id: The shell id on the remote machine.
+         See #open_shell
+        @param string command_id: The command id on the remote machine.
+         See #run_command
+        @param string stdin_input: The input unicode string to be sent.
+        @return: None
+        """
+        if isinstance(stdin_input, str):
```

Comment:
Looking at the documentation, the following should be a decent work around:
```
stdin_input = text_type(stdin_input).encode("UTF-8")
```
Do you agree?

Reply:
Testing with python 2.7:
```
>>> isinstance("test", six.text_type)
False
>>> six.text_type("test").encode("UTF-8")
'test'
>>> six.text_type("test")
u'test'
```

Reply:
> Looking at the documentation, the following should be a decent work around:

Now you are potentially converting a byte string to a text string then back to bytes which will cause all sorts of issues and weird conversions. You should take bytes as they are and just base64 encode them directly and only encode text strings.

It also has the same issue when you try to convert a byte string with non-ascii chars like `é` as `text_type("str")` will traceback on Python 2.

Ultimately what I think you want is to accept both byte and text strings like.

```
if isinstance(stdin_input, six.text_type):
    stdin_input = stdin_input.encode("TBD")  # This still needs to be solved

stdin_envelope['#text'] = base64.b64encode(stdin_input)
```

This allows you to do the following;


```
# Python 2
send_command_input(shell_id, cmd_id, b"café")
send_command_input(shell_id, cmd_id, "café")
send_command_input(shell_id, cmd_id, u"café")

# Python 3
send_command_input(shell_id, cmd_id, b"café")
send_command_input(shell_id, cmd_id, "café")
send_command_input(shell_id, cmd_id, u"café")
```

Reply:
> Looking at the documentation, the following should be a decent work around:

Now you are potentially converting a byte string to a text string then back to bytes which will cause all sorts of issues and weird conversions. You should take bytes as they are and just base64 encode them directly and only encode text strings.

It also has the same issue when you try to convert a byte string with non-ascii chars like `é` as `text_type("str")` will traceback on Python 2.

Ultimately what I think you want is to accept both byte and text strings like.

```
if isinstance(stdin_input, six.text_type):
    stdin_input = stdin_input.encode("TBD")  # This still needs to be solved

stdin_envelope['#text'] = base64.b64encode(stdin_input)
```

This allows you to do the following and all examples will results in the payload of `Y2Fmw6k=` (if we keep the default of utf-8 encoding);


```
# Python 2
send_command_input(shell_id, cmd_id, b"café")
send_command_input(shell_id, cmd_id, "café")  # On Python 2 this is the same as the above
send_command_input(shell_id, cmd_id, u"café")


# Python 3
send_command_input(shell_id, cmd_id, b"caf\xc3\xa9")
send_command_input(shell_id, cmd_id, "café")  # On Python 3 this is the same as below
send_command_input(shell_id, cmd_id, u"café")
```

So if a user wanted to pass a text then they should prefix their string with `u""` to ensure compatibility across both Python versions. But still having the ability to pass a byte string if they really wanted to. This would enable me to do something like the below which allows me to specify the encoding I want without relying on the defaults in the function.

```
send_command_input(shell_id, cmd_id, u"café".encode('windows-1252'))
```

Allowing a byte string input also allows me to send actual bytes that may not relate to a character code point if that's what my process expects.

Reply:
Well, what do you propose we do about this section of code to maximize compatibility while still maintaining the ability to pass both unicode strings as well as byte strings into the function?

Reply:
Exactly what I posted

Reply:
Changed to:
```
if isinstance(stdin_input, text_type):
    stdin_input = stdin_input.encode("UTF-8")
```

Reply:
That solves the input type problem but we still need to figure out whether `utf-8` is the right encoding to use when encoding a text string. I went into detail around why this can be problematic, especially since the default codepage is `437` which is not `utf-8`. Curious as to what your thoughts are alongside @badcure and @nitzmahone.

Reply:
Just to show you what I mean, consider running

```
import base64

from winrm import Protocol


def set_content(p, shell_id, path, content, encoding):
    script = "; ".join([
        "$path = '%s'" % path,
        # The default encoding of Set-Content is the System code page (not the console code page) which on my host is
        # windows-1252 "ANSI". This is important as we use that encoding scheme to decode the bytes read
        "Set-Content -Path $path -Value ($input | Out-String)",
        "[System.Convert]::ToBase64String([System.IO.File]::ReadAllBytes($path))"
    ])
    cmd_id = p.run_command(shell_id, "powershell", ["\"%s\"" % script])
    p.send_command_input(shell_id, cmd_id, content.encode(encoding))
    out = p.get_command_output(shell_id, cmd_id)
    p.cleanup_command(shell_id, cmd_id)

    # The files as saved in ANSI (windows-1252) so we use that encoding to decode the raw bytes and get our text value
    return base64.b64decode(out[0].decode('utf-8').rstrip()).decode('windows-1252').rstrip()


p = Protocol("http://server:5985/wsman", username="username", password="password")
shell_id = p.open_shell()

print("UTF-8: %s" % set_content(p, shell_id, "C:\\temp\\output-utf-8.txt", u"café", "utf-8"))
print("437: %s" % set_content(p, shell_id, "C:\\temp\\output-437.txt", u"café", "437"))

p.close_shell(shell_id)
```

This will create 2 files `C:\temp\output-utf-8.txt` and `C:\temp\output-437.txt`. The first example I an encoding the string using `utf-8` (like the default in this function) and the 2nd I am encoding it according to my console's set codepage. Running this will output

```
UTF-8: caf+¬
437: café
```

Obviously the 2nd one is correct and what we expect which shows us the default may not be optimal. If I was to view that file on the Windows host in notepad I would see the same thing

Reply:
That really depends on what encoding your notepad uses as the default for opening the file as determined by your geographic location.. As to why I sided with UTF-8. The following should explain it:
![image](https://user-images.githubusercontent.com/4672283/62912920-4c2f3280-bd47-11e9-8f0d-174cc7e37a21.png)
And the reason for this is Linux, regardless of region, usually uses "UTF-8".  That being said, after doing some more research on page 437 encoding, it is the default encoding used for older versions of DOS, but newer versions of windows default to "UTF-16" but is region dependent.
Why not use code page 1252? 
https://en.wikipedia.org/wiki/Windows-1252
But really, it should be more of a question of what the windows remote management service is expecting before passing characters on to perspective programs.  I'm going down for tonight, but can you run a test with the code in question to interact with a running process to echo out different characters using UTF-8 encoding to make sure that when they are echoed out, that they are not mangled.

Reply:
>  As to why I sided with UTF-8. The following should explain it

That's all well and good for the Internet but Windows isn't really utf-8 friendly when it comes to the console. It's getting a lot better but it's still not there yet.

> That being said, after doing some more research on page 437 encoding, it is the default encoding used for older versions of DOS, but newer versions of windows default to "UTF-16" but is region dependent

You are confusing the encodings that are used throughout Windows, there are multiple locations/settings where you can set a the culture/format/codepage/encoding you name it but we are talking about console encoding code pages. The Win32 API has 2 variants a *A (ANSI) and *W (Wide) where the Wide is a UTF-16 like encoding but console encoding is a whole separate beast. The default value is dependent on the region of Windows yes but in the case of pywinrm we do set [437](https://github.com/diyan/pywinrm/blob/master/winrm/protocol.py#L113) as the default and this is set as the spawned processes console's code page regardless of its actual region format.

> Why not use code page 1252?

Because that's not what we use when we create the shell by default.

> But really, it should be more of a question of what the windows remote management service is expecting before passing characters on to perspective programs

It's dependent on the console code page set up.

> I'm going down for tonight, but can you run a test with the code in question to interact with a running process to echo out different characters using UTF-8 encoding to make sure that when they are echoed out, that they are not mangled.

Sure run the following (Python 3 is recommended as you can see the byte string):


```
import base64

from winrm import Protocol


def echo_input(p, content, encoding, **kwargs):
    shell_id = p.open_shell(**kwargs)
    script = u"""
$win_string = "{0}`r`n"  # The newlines are added so the length matches the input string (which have newlines)
$input_string = $input | Out-String
"From Windows: '$win_string', Length: $($win_string.Length)"
"From Input: '$input_string', Length: $($input_string.Length)"
""".format(content).encode('utf-16-le')

    cmd_id = p.run_command(shell_id, "powershell", ["-EncodedCommand", base64.b64encode(script).decode('utf-8')])
    p.send_command_input(shell_id, cmd_id, content.encode(encoding), end=True)
    out = p.get_command_output(shell_id, cmd_id)
    p.cleanup_command(shell_id, cmd_id)
    p.close_shell(shell_id)

    return out[0]


p = Protocol("http://server:5985/wsman", username="user", password="pass")

print(echo_input(p, u"café", "utf-8"))
print(echo_input(p, u"café", "437"))
print(echo_input(p, u"café", "utf-8", codepage=65001))
print(echo_input(p, u"café", "437", codepage=65001))
```

It will output the following:

```
b"From Windows: 'caf\x82\r\n', Length: 6\r\nFrom Input: 'caf\xc3\xa9\r\n', Length: 7\r\n"
b"From Windows: 'caf\x82\r\n', Length: 6\r\nFrom Input: 'caf\x82\r\n', Length: 6\r\n"
b"From Windows: 'caf\xc3\xa9\r\n', Length: 6\r\nFrom Input: 'caf\xc3\xa9\r\n', Length: 6\r\n"
b"From Windows: 'caf\xc3\xa9\r\n', Length: 6\r\nFrom Input: 'caf\xef\xbf\xbd\r\n', Length: 6\r\n"
```

We can see in the first 2 outputs where we don't set a codepage that when Windows outputs the string `café` it looks like `caf\x82\r\n' that's 6 characters long. If we compare that to the same string we send on the input but encoded as `utf-8` we can see that it is 7 characters long and now the value is `caf\xc3\xa9\r\n`. While this is a legitimate utf-8 string repr of `café` the Windows console does not see it that way. It see it as a 7 character long string where the input bytes were encoded using the console's codepage, e.g. `caf├⌐`.

Now you could be thinking that's fine I will just decode the output as a `utf-8` string but try to decode that output using utf-8 and you will raise an exception.

```
b"From Windows: 'caf\x82\r\n', Length: 6\r\nFrom Input: 'caf\xc3\xa9\r\n', Length: 7\r\n".decode('utf-8')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x82 in position 18: invalid start byte
```

This is because the output bytes are encoded based on the codepage set (437) and not just what the user wants.

Finally the last 2 examples show what happens in those same scenarios but if we have the codepage set to `65001` (utf-8). We can see that out input encoding of `utf-8` is good as it matches out console input but if we were to use `437` we get an unknown character at the end `caf�`.


So putting all this aside I think we have the following choices;

* Keep the default of `utf-8` as the encoding when the input is a text string
    * We can just document this and if someone wants a particular encoding that is not `utf-8` they can just pass a byte string with whatever they need
* Align the default to `437` to match our default when we create the shell
    * This keeps things aligned but doesn't help people who have changed their codepage, but at least they are aware of the encoding side if they touched that kwarg unlike in 1 where people are oblivious and just expect it to work
* Add a kwarg to `send_command_input` where you can control the encoding
    * This is somewhat useful but may not be needed as someone can just pass in a byte string that is already encoded the way they want
* Keep track of the codepage used to create a shell based on the shell id and use that for encoding
    * I don't think this is feasible as not all code pages in Windows actually matches up with Python encoding values, e.g. `.encode("65001")` doesn't work

Personally I would go with option 1 or 2 maybe leaning towards 2 based on how our default shell codepage is `437` but I'm bringing this up so can all be aware of what is going on and have a group discussion and where to go from here.

Reply:
I've aligned it to the default 437 as you desire  If it ends up being the wrong choice, which doesn't seem likely given the above results, we can always change it to something else later.

Reply:
I'm going to resolve this issue unless there is anything else.  Is there anything else?